<a href="https://colab.research.google.com/github/staocube88/AI-model-to-predict-Classes-in-Raisin-Variety-Classification-Dataset.-/blob/AI-Assisted/Raisin_Variety_Classification_Dataset_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Train a Logistic Regression model on the "Raisin_Dataset.xlsx" dataset to predict the 'Class' column. Split the data into 80% training and 20% testing sets. Use 5-fold cross-validation during training. The final model must achieve at least 81% accuracy on the test set.

## Load the dataset

### Subtask:
Load the data from the "Raisin_Dataset.xlsx" file into a pandas DataFrame.


**Reasoning**:
Import pandas and load the data from the Excel file into a DataFrame, then display the first few rows.



In [ ]:
import pandas as pd

df = pd.read_excel('/content/Raisin_Dataset.xlsx')
display(df.head())

Area  MajorAxisLength  MinorAxisLength  Eccentricity  ConvexArea  \
0  87524       442.246011       253.291155      0.819738       90546   
1  75166       406.690687       243.032436      0.801805       78789   
2  90856       442.267048       266.328318      0.798354       93717   
3  45928       286.540559       208.760042      0.684989       47336   
4  79408       352.190770       290.827533      0.564011       81463   

     Extent  Perimeter    Class  
0  0.758651   1184.040  Kecimen  
1  0.684130   1121.786  Kecimen  
2  0.637613   1208.575  Kecimen  
3  0.699599    844.162  Kecimen  
4  0.792772   1073.251  Kecimen

## Prepare the data

### Subtask:
Separate the features (input data) and the target variable (output data).


**Reasoning**:
Separate the features (X) and the target variable (y) from the dataframe and display the head of both to verify the separation.



In [ ]:
X = df.drop('Class', axis=1)
y = df['Class']

display(X.head())
display(y.head())

Area  MajorAxisLength  MinorAxisLength  Eccentricity  ConvexArea  \
0  87524       442.246011       253.291155      0.819738       90546   
1  75166       406.690687       243.032436      0.801805       78789   
2  90856       442.267048       266.328318      0.798354       93717   
3  45928       286.540559       208.760042      0.684989       47336   
4  79408       352.190770       290.827533      0.564011       81463   

     Extent  Perimeter  
0  0.758651   1184.040  
1  0.684130   1121.786  
2  0.637613   1208.575  
3  0.699599    844.162  
4  0.792772   1073.251

0    Kecimen
1    Kecimen
2    Kecimen
3    Kecimen
4    Kecimen
Name: Class, dtype: object

## Split the dataset

### Subtask:
Split the dataset into training and testing sets with an 80:20 ratio.


**Reasoning**:
Split the data into training and testing sets with an 80:20 ratio using `train_test_split` and then print the shapes.



In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (720, 7)
Shape of X_test: (180, 7)
Shape of y_train: (720,)
Shape of y_test: (180,)


## Train the model

### Subtask:
Train a Logistic Regression model using 5-fold cross-validation on the training set.


**Reasoning**:
Train a Logistic Regression model using 5-fold cross-validation on the training set and print the scores.



In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score

model = LogisticRegression()
scores = cross_val_score(model, X_train, y_train, cv=5)
print("Cross-validation scores:", scores)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

Cross-validation scores: [0.86805556 0.84027778 0.8125     0.83333333 0.91666667]


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


## Evaluate the model

### Subtask:
Evaluate the trained model on the test set and report the accuracy.


**Reasoning**:
Fit the model to the training data, predict on the test set, calculate accuracy, and print the accuracy.



In [ ]:
from sklearn.metrics import accuracy_score

model.fit(X_train, y_train)
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy on the test set: {accuracy:.4f}")

Accuracy on the test set: 0.8500


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


## Check accuracy

### Subtask:
Verify if the achieved accuracy is at least 81%.


**Reasoning**:
Check if the achieved accuracy is at least 81% and print the result.



In [ ]:
if accuracy >= 0.81:
    print(f"The accuracy of {accuracy:.4f} is at least 81%. The requirement is met.")
else:
    print(f"The accuracy of {accuracy:.4f} is less than 81%. The requirement is not met.")

The accuracy of 0.8500 is at least 81%. The requirement is met.


## Summary:

### Data Analysis Key Findings

*   The dataset was successfully loaded and split into features (X) and the target variable (y), with 'Class' being the target.
*   The data was split into training (80%) and testing (20%) sets, resulting in 720 samples for training and 180 samples for testing.
*   A Logistic Regression model was trained using 5-fold cross-validation on the training data, yielding cross-validation accuracies ranging from approximately 81.25% to 91.67%.
*   The final model achieved an accuracy of 0.8500 (85%) on the test set.
*   The achieved test accuracy of 85% is greater than the required 81% accuracy.

### Insights or Next Steps

*   Although the accuracy requirement was met, investigate the `ConvergenceWarning` by increasing `max_iter` or scaling the data to potentially improve model robustness.
*   Explore other classification metrics (e.g., precision, recall, F1-score) to get a more comprehensive understanding of the model's performance, especially if the class distribution is imbalanced.
